File: packages/python-sdk/src/blindfold/regex/detectors/us/zip_code.py
"""US ZIP Code detector (context-required)"""

import re

from ...base import RegexDetector
from ...entities import EntityType
from ...registry import register_region


def _zip_valid(text: str) -> bool:
    """Reject ZIPs starting with 000 (no valid US ZIP starts with 000)."""
    digits = "".join(c for c in text if c.isdigit())
    return not digits.startswith("000")


@register_region("us")
class ZipCodeDetector(RegexDetector):
    entity_type = EntityType.ZIP_CODE
    score = 0.70
    context_keywords = [
        "zip", "postal", "zip code", "zipcode", "postal code",
        "address", "city", "state", "mailing", "shipping",
        "billing", "delivery", "location", "resident", "zip+4",
        "usa", "united states",
    ]
    context_required = True
    context_window = 150

    # 5 digits or 5+4 format
    pattern = re.compile(r"\b\d{5}(?:-\d{4})?\b")
    validator = staticmethod(_zip_valid)

File: packages/js-sdk/src/regex/detectors/us/zip-code.ts
// US ZIP Code detector (context-required)

import { RegexDetector } from '../../base'
import { EntityType } from '../../entities'
import { registerRegion } from '../../registry'

function zipValid(text: string): boolean {
  let digits = ''
  for (const c of text) {
    if (c >= '0' && c <= '9') digits += c
  }
  return !digits.startsWith('000')
}

class ZipCodeDetector extends RegexDetector {
  entityType = EntityType.ZIP_CODE
  score = 0.7
  contextKeywords = [
    'zip', 'postal', 'zip code', 'zipcode', 'postal code',
    'address', 'city', 'state', 'mailing', 'shipping',
    'billing', 'delivery', 'location', 'resident', 'zip+4',
    'usa', 'united states',
  ]
  contextRequired = true
  contextWindow = 150

  // 5 digits or 5+4 format
  pattern = /\b\d{5}(?:-\d{4})?\b/g
  validator = zipValid
}

registerRegion('us', ZipCodeDetector)

export { ZipCodeDetector }
